### PR TITLE
[WIP] Extend GasFeeController to poll for network status

### PR DESCRIPTION
**Pull request by mcmire**
When a user is about to send a transaction or is looking at a swap
quote, we would like to inform them if the network is busy so that we
can push them to use a lower fee setting. GasFeeController already
provides a way to poll for fee estimates, which we employ on transaction
preview screens. This commit updates the polling code so that we also
gauge network status as we pull estimates. This is done by hitting
another endpoint in the MetaSwap API which specifically gives us a base
fee threshold we can use to determine whether the network is busy
(falling back to using `eth_feeHistory` if the API is not available).

---

References https://github.com/MetaMask/metamask-extension/issues/11307.

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -1161,6 +1161,13 @@
   dependencies:
     "@sinonjs/commons" "^1.7.0"
 
+"@sinonjs/fake-timers@^8.0.1":
```
Comment (wachunei) [inline yarn.lock:1164]: Q: Is there a particular feature this dependency has that made you choose it over jest's fake timers?

Comment (mcmire) [reply yarn.lock:1164]: Good question. The functionality that is being tested inside of GasFeeController involves asynchronous behavior via not only `setTimeout` but also via promises. That is, inside of the `setTimeout`, we run one or more promises, wait for all to finish, then schedule the next `setTimeout`, then run more promises, then schedule the next `setTimeout`, etc. I tried Jest's fake timers at first and ran into some perplexing timing issues. Although Jest was effective in stubbing out `setTimeout`, it seemed that the promises "leaked out" of the setTimeout and were therefore running after key assertion lines in the tests ran. This made it impossible to truly test the behavior in question. I did some research and it seems that Jest's fake timers have had [issues](https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises) in the past with this sort of thing. In Jest 26+, you can supposedly say `jest.useFakeTimers('modern')` to switch over to a different implementation [which makes use of Sinon's fake timers](https://github.com/facebook/jest/pull/7776), but I tried this and it seemed to not know what I was talking about. Plus, this "modern" implementation [seems to have some quirks too](https://github.com/facebook/jest/issues/10221#issuecomment-654687396) that seems to indicate that Jest makes some modifications to Sinon's timers. In the end, using Sinon directly solved the problem.

Comment (wachunei) [reply yarn.lock:1164]: This is great to know, thank you for the answer!

File: src/gas/GasFeeController.ts [hunk 2]
```diff
@@ -1,10 +1,9 @@
 import type { Patch } from 'immer';
```
Comment (mcmire) [inline src/gas/GasFeeController.ts:1]: This seems like a lot, so I'll try to summarize:

GasFeeController was designed for polling for gas fee estimates. These changes extend the polling logic such that it is now also possible to poll for network congestion. We achieve this by removing the concept of poll tokens and replacing this with known pollable items ("gasFeeEstimates" and "networkCongestion") as well as a poll queue (a mapping of pollable items to functions).

One could argue that these changes do not belong in this controller, and one *may* be correct. However, I tried to put the bits about network congestion / fee history in a separate controller, and the problem is that if we want to poll for two things at the same time, that makes that really hard to do. It seems that we might could benefit from a PollingController, but I really wasn't sure what the best approach here, so it seemed easier to extend this controller. Let me know if you have any alternative opinions though!

File: src/gas/GasFeeController.ts [hunk 3]
```diff
@@ -426,48 +482,53 @@ export class GasFeeController extends BaseController<
...
+   */
   stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
-    }
-    this.pollTokens.clear();
-    this.resetState();
+    this.pausePolling();
+    this.pollQueue.clear();
+    // XXX: Previously this called resetState(), do we still want that?
```
Comment (mcmire) [inline src/gas/GasFeeController.ts:531]: I really tried to keep the existing behavior with these changes and not introduce backward incompatibilities, but this is one case where I opted to break that. Previously when calling `stopPolling`, we cleared the entire state. We can't do that now since not everything in state has to do strictly with gas fee estimates. What do we want to do here?

File: src/gas/GasFeeController.ts [hunk 4]
```diff
@@ -522,6 +579,69 @@ export class GasFeeController extends BaseController<
...
+      const sortedPriorityFees: number[] = feeHistory.blocks
+        .map((block) => block.priorityFeesByPercentile.get(50))
+        .filter(isNumber)
+        .sort((a, b) => a - b);
+
+      if (sortedPriorityFees.length > 0) {
+        const minPriorityFee = sortedPriorityFees[0];
+        const maxPriorityFee =
+          sortedPriorityFees[sortedPriorityFees.length - 1];
+        isNetworkCongested = maxPriorityFee >= minPriorityFee * 1.1;
```
Comment (mcmire) [inline src/gas/GasFeeController.ts:633]: This is the algorithm I chose to determine whether the network is congested, but please note that I haven't run this by anything yet — I just went with something to get this PR started. That's why this is still a draft. There are ongoing discussions about this in Slack, so stay tuned.

File: src/gas/fetchFeeHistory.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * @param args.percentiles - A tuple of numbers between 1 and 100 which will advise how
+ * priorityFeesByPercentile in the return value will be formed. When Ethereum runs the
+ * eth_feeHistory method, for each block it is considering, it will calculate the ratio of gas used
+ * for each transaction compared to the total gas used for the block, then sort all transactions in
+ * the block by this ratio. For each percentile given, it will then find the transaction whose gas
+ * used ratio matches the percentile and capture that transaction's priority fee. Hence,
+ * priorityFeesByPercentile represents the priority fees of transactions at key gas used ratios.
+ * @returns The fee history data.
+ */
+export async function fetchFeeHistory({
```
Comment (mcmire) [inline src/gas/fetchFeeHistory.ts:85]: This function works different than the existing function in `util.ts` for fetching gas fee estimates in that instead of taking a URL it takes an EthQuery. We may change this to hit a Metaswap API instead of Infura, so this is subject to change. I just went with something for now.

File: src/gas/fetchFeeHistory.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  };
+};
+
+/**
+ * Converts a hexadecimal string to a decimal number.
+ *
+ * @param hex - A string encoding a hexadecimal number (with a "0x") prefix.
+ * @returns The number in decimal.
+ */
+function hexToDec(hex: string): number {
```
Comment (mcmire) [inline src/gas/fetchFeeHistory.ts:60]: I know that we already have some convenience functions around working with hex values in a couple of different places, but it didn't seem like they did exactly what I wanted. For instance, `weiHexToGweiDec` in `util.ts` converts WEI to GWEI, and I don't think we want that in this case. `normalizeGWEIDecimalNumbers` in `gas-util.ts` is better but it ends up converting the final result to a string, and seems like that is unnecessary (if we really wanted a big number, wouldn't we use BN?).

we convert a hex string, we still end up with a string. I really wonder why we are doing this. It seems that if there is a reason for doing so, say, to prevent floating point issues, then BN or BigNumber would be useful in representing those numbers. Otherwise, it does not

File: .eslintrc.js [hunk 7]
```diff
@@ -65,6 +103,10 @@ module.exports = {
     'no-param-reassign': 'off',
     radix: 'off',
     'require-atomic-updates': 'off',
+    'jsdoc/match-description': [
```
Comment (mcmire) [inline .eslintrc.js:106]: Another thing that's being added temporarily to allow for a list of links at the end of a description block. Will also move this to a separate PR, but here for now to get the linter step to pass.

File: .eslintrc.js [hunk 8]
```diff
@@ -14,6 +14,44 @@ module.exports = {
     {
       files: ['*.test.ts', '*.test.js'],
       extends: ['@metamask/eslint-config-jest'],
+      rules: {
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply .eslintrc.js:17]: I'm guessing your motivation here is to avoid writing doc comments for test helpers. Could you elaborate on why you want to avoid doing that? I find such comments useful for understanding how to use the test helpers, and for understanding what the tests do.

Also, I suspect that `jsdoc/require-jsdoc` is the only rule that needs to be disabled here.  The rest of the rules are for ensuring any JSDoc comments that do exist are complete and formatted correctly. Even if we want to make JSDoc comments optional in tests, we'd still want them to be complete and correctly formatted.

[the remark this answers is not in the document]
Comment (mcmire) [reply .eslintrc.js:17]: So as a little context, I have some helper functions I've added to the tests for GasFeeController. I realize this is a little different from the pattern set by other test files, but for these tests it seemed like a better way to organize things as we have a lot of setup code in this file. I am curious whether we should require people to write documentation for functions that are local to test files. In my view, good documentation explains the "why" behind the code, not the "what", and so the best documentation exists in places where the "why" is not obvious. I may be a little biased, but I think the context/intent behind the functions in `GasFeeController.test.ts` is fairly obvious provided that a good name is being provided for those functions and once you see how they are being used, and it would be difficult to provide more information about such functions that could not already be easily derived from the code. For instance, what could you write about `buildReturnValueForFetchGasEstimates`? If you don't know how GasFeeController works, you'd have to explain how it works (but you can do that elsewhere). If you _do_ know how it works, you can see where it's being used in the test, which is as mock data for a `fetchGasEstimates` function being passed into the controller. If that is difficult to do, then there are probably readability issues with the entire file anyway.

Based on this, my worry is that by requiring documentation across the board in test files, we are adding an obstacle to writing tests. That said, perhaps disabling these ESLint rules is a hasty decision given that we don't have many test files that keep their own helpers. So on that alone I would be willing to take these changes back and wait until we have more of these types of situations.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply .eslintrc.js:17]: Well, the rules only require documentation for helper functions, not the tests themselves. And it is important that helper functions are clearly understood if the tests are to remain readable. Honestly if these modest documentation requirements discouraged writing helper methods completely that wouldn't be a huge loss for me, because they have a tendency to make tests difficult to understand anyway. Better to have long, verbose tests where all of the setup is laid bare, than one involving an idiosyncratic mock that was created by a "createMockX" function that could be doing god knows what.

Honestly `buildReturnValueForFetchGasEstimates` does seem like an example of a difficult-to-understand helper function (at least devoid of further context - I haven't read this PR yet). What does this mock represent? Why were the specific values used? What does this tell us about the code under test? Does this mock represent an edge case, a base case, a typical one, or?

A short description like "Return an example gas estimate API response featuring a typical distribution of estimate times" would help me a lot as a reader. It would tell me that it was meant to represent a typical gas estimate, that I'd have to think about edge cases differently (e.g. what if the API returned zero? What if the low estimate was higher than the high one? What if all of the estimates were very high or very low, with little variation? What if an error was returned?), and maybe add tests for those or add validation to prevent those cases in the API module.

All of that might be clear from a careful read, but a helper function that has to be read carefully to be understood will not improve readability. WIth a nice concise description, I could hover over the helper method in VSCode and see the description without having to leave the test I'm reading. As opposed to the alternative of switching contexts, carefully analyzing what the mock does and trying to guess why, then trying to keep that all in my head when I switch back and try it recall what the test I was reading was doing.

File: src/gas/GasFeeController.ts (hunk 2 above)
Comment (adonesky1) [reply src/gas/GasFeeController.ts:1]: Haven't read through the changes here yet but, in theory, including polling for 'networkCongestion' in this controller makes sense to me (especially given, as you point out, we intend to poll the same RPC method as an estimate fee estimate fallback.

File: .eslintrc.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (rekmarks) [reply .eslintrc.js:17]: I agree with @mcmire that this is not necessarily the forum for litigating the JSDoc lint rules, but since we are here, I'll put my stake in the ground, too.

Having written some awful test helpers in my day, I have to agree with @Gudahtt here. Writing a test helper function docstring should never take more than a few minutes at most, and with autocomplete in e.g. VS Code (mac: cmd+shift+enter after typing `/** */` immediately above a function or method) it's damn-near trivial.

In general, we are asking whether it's worth maybe 2-10 more minutes per function to make all of our code — tests and otherwise — more readable, for everyone, forever. I wrote a bunch of docstrings for a large class earlier today and I hated every minute of it, but I can't say that it's not worth it. We'll get the time I spent writing them back twelve times over in the next 3-6 months, to say nothing of the long term.

[the remark this answers is not in the document]
Comment (mcmire) [reply .eslintrc.js:17]: Okay thanks for the feedback on this! I'll remove these changes and add a description for this function.

File: src/gas/GasFeeController.ts (hunk 4 above)
Comment (mcmire) [reply src/gas/GasFeeController.ts:622]: Most recent discussion is here: https://consensys.slack.com/archives/C021UN08FBJ/p1632245722147100

File: src/gas/GasFeeController.test.ts [hunk 9]
```diff
@@ -1,15 +1,25 @@
-import { stub } from 'sinon';
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/gas/GasFeeController.test.ts:1]: Sounds great! Definitely agree that we should avoid using `nock` here. That was the whole point of making the `fetch_` methods constructor parameters in the first place.

Could these test refactors be submitted as a separate PR? That would make this a lot easier to review. These changes alone seem pretty significant.

[the remark this answers is not in the document]
Comment (mcmire) [reply src/gas/GasFeeController.test.ts:1]: Yup I can certainly do that!

File: src/gas/GasFeeController.ts [hunk 10]
```diff
@@ -309,56 +355,61 @@ export class GasFeeController extends BaseController<
...
-      await this._fetchGasFeeEstimateData();
-      this._poll();
+  /**
+   * Makes a request for the given item, updating the state internally with the resulting data, then
+   * adds the request to the polling queue. The polling timer is then started if it is not already
+   * started.
+   *
+   * @param item - The known name of a piece of data for which we are interested in polling.
+   */
+  async updateWithAndStartPollingFor(item: keyof PollableItems): Promise<void> {
```
Comment (mcmire) [inline src/gas/GasFeeController.ts:407]: These changes remove the concept of a polling token, but on second thought, we should actually keep that. In rooting around in the `extension` repo, I've recently learned that we start duplicate instances of polling for gas fee estimates in multiple places. We shouldn't be doing this, but we do. So by no longer keeping track of polling tokens, we will make it so that if a component stops polling for gas fee estimates it kills polling for ALL instances.

File: src/gas/GasFeeController.ts (hunk 2 above)
Comment (mcmire) [reply src/gas/GasFeeController.ts:1]: Resolving this conversation since I ended up coming up with a simpler solution.

File: src/gas/GasFeeController.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply src/gas/GasFeeController.test.ts]: Resolving this conversation since this branch was rebased on top of other work.

File: src/gas/calculateBusyThreshold.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import { BN } from 'ethereumjs-util';
```
Comment (mcmire) [inline src/gas/calculateBusyThreshold.ts:1]: See https://gitlab.com/ConsenSys/codefi/products/metaswap/gas-api/-/blob/017436f628b2d5967f6e8734780a9114f9e58af9/src/server.ts#L186 and https://gitlab.com/ConsenSys/codefi/products/metaswap/gas-api/-/blob/017436f628b2d5967f6e8734780a9114f9e58af9/src/eip1559/feeHistory.ts#L152 for the original source.

**Comment by mcmire** [comment]
This may end up being obsoleted by https://github.com/MetaMask/controllers/pull/632. I'll leave this open until that's merged.

**Comment by mcmire** [comment]
Closing since we can now use `networkCongestion` to gauge busyness.
